packages/http-crawler/src/internals/file-download.ts

```diff
@@ -0,0 +1,217 @@
+import { finished } from 'stream/promises';
+import { isPromise } from 'util/types';
+
+import type { Dictionary } from '@crawlee/types';
+
+import type {
+    ErrorHandler,
+    GetUserDataFromRequest,
+    HttpCrawlerOptions,
+    InternalHttpCrawlingContext,
+    InternalHttpHook,
+    RequestHandler,
+    RouterRoutes,
+} from '../index';
+import {
+    HttpCrawler,
+    Router,
+} from '../index';
+
+export type FileDownloadErrorHandler<
+    UserData extends Dictionary = any, // with default to Dictionary we cant use a typed router in untyped crawler
+    JSONData extends Dictionary = any, // with default to Dictionary we cant use a typed router in untyped crawler
+    > = ErrorHandler<FileDownloadCrawlingContext<UserData, JSONData>>;
+
+export type StreamHandlerContext = Omit<FileDownloadCrawlingContext, 'body' | 'response' | 'parseWithCheerio' | 'json' | 'addRequests' | 'contentType'> & {
+    stream: ReadableStream;
+};
+
+type StreamHandler = (context: StreamHandlerContext) => void | Promise<void>;
+
+export type FileDownloadCrawlerOptions<
+UserData extends Dictionary = any, // with default to Dictionary we cant use a typed router in untyped crawler
+JSONData extends Dictionary = any, // with default to Dictionary we cant use a typed router in untyped crawler
+> =
+(Omit<HttpCrawlerOptions<FileDownloadCrawlingContext<UserData, JSONData>>, 'requestHandler' > & { requestHandler?: never; streamHandler?: StreamHandler }) |
+// eslint-disable-next-line max-len
+(Omit<HttpCrawlerOptions<FileDownloadCrawlingContext<UserData, JSONData>>, 'requestHandler' > & { requestHandler: FileDownloadRequestHandler; streamHandler?: never });
```

Comment:
this should be also indented i guess? overall the indents are very random in this file :D